File: src/main/java/org/khun/leetcode/SlidingWindows/LongestSubStrWithoutRepeatingChar.java
package org.khun.leetcode.SlidingWindows;

import java.util.HashSet;

public class LongestSubStrWithoutRepeatingChar {
    public static int lengthOfLongestSubstring(String s) {
        if (s == null || s.isEmpty() || s.isBlank()) return 0;

        HashSet<Character> charSet = new HashSet<>();
        int maxLenght = 0;
        int left = 0;

        for (int right = 0; right<s.length(); right ++){
            while (charSet.contains(s.charAt(right))){
                charSet.remove(s.charAt(left));
                left++;
            }
            charSet.add(s.charAt(right));
            maxLenght = Math.max(maxLenght, charSet.size());
        }
        return maxLenght;
    }

    public static void main(String[] args) {
        System.out.println(lengthOfLongestSubstring("abcabcbb"));
        System.out.println(lengthOfLongestSubstring("bbbbb"));
        System.out.println(lengthOfLongestSubstring("pwwkew"));
    }
}

File: src/main/java/org/khun/leetcode/SlidingWindows/LongestSubStrWithoutRepeatingChar.kt
package org.khun.leetcode.SlidingWindows

/**
 * LeetCode 3. Longest Substring Without Repeating Characters
 */
fun lengthOfLongestSubstring(s: String): Int {
    val seen = mutableSetOf<Char>()
    var index = 0
    var len = 0
    for (ch in s) {
        while (seen.contains(ch)) {
            seen.remove(s[index++])
        }
        seen.add(ch)
        len = maxOf(len, seen.size)
    }
    return len
}

fun main() {
    val s1 = "abcabcbb"
    println(lengthOfLongestSubstring(s1))

    val s2 = "bbbbb"
    println(lengthOfLongestSubstring(s2))

    val s3 = "pwwkew"
    println(lengthOfLongestSubstring(s3))
}